### PR TITLE
Add route overview button and animation to NavigationView

### LINKED ISSUE #950: Add route overview button
We should add a route overview button to the navigation UI. Zooming out to see an overview of the full route and then zooming back in is a very common action when using turn-by-turn navigation. Adding this feature makes it a lot safer for a user to do this while driving.

Here's what it looks like on iOS:

cc @mapbox/navigation-android

### PR DESCRIPTION
**Pull request by danesfeder**
Closes #950 

TODO:
- [x] Add button with chosen asset from @Cjballard 
- [x] Show resume button after overview clicked
- [x] Add test for dynamic camera providing list of `Point`

Animation running wired up to the `X` button:
ezgif com-video-to-gif 1

cc @brsbl

### CODE REVIEW

**Comment by Cjballard** [comment]
[routepreview_icons.zip](https://github.com/mapbox/mapbox-navigation-android/files/2036876/routepreview_icons.zip)

Attaching light and dark icons for route preview. @danesfeder

**Comment by brsbl** [comment]
Love the dynamic camera.

@danesfeder is it easy enough to adjust the spacing between the sound and feedback icons in this PR as well so they better match iOS? the thinking being grouping things that obscure the map together makes them less intrusive overall -- thoughts @Cjballard? 

also:

- [ ] Cut a ticket in /mapbox-navigation-ios to mirror this behavior 🙂

**Comment by danesfeder** [comment]
@brsbl sure thing - would like @Cjballard 👀 on this as well. I have them spaced a little more to create larger touch points.  Hitting that feedback button accidentally when you're trying to mute is pretty jarring

**Comment by Cjballard** [comment]
@brsbl @danesfeder Agreed. I think there's a middle ground we can hit that still reserves those touchpoints while grouping them more tightly and revealing more of the map. Maybe 10-20dp closer together could work.

**Comment by danesfeder** [comment]
@Cjballard Here's what the current setup looks like, let me know what you think / want to tweak 👍 

### `Portrait`
ezgif com-video-to-gif

### `Landscape`
ezgif com-video-to-gif 1

**Comment by Cjballard** [comment]
@danesfeder Looks good to me! Only adjustment I might make is to make the separating line a bit shorter, closer to the height (but still a little taller) than the route preview icon: 

tt-example

**Comment by danesfeder** [comment]
@Cjballard Updated:
device-2018-06-05-164139

device-2018-06-05-164333

**Comment by Cjballard** [comment]
@danesfeder Looks good to me 👍

**Comment by danesfeder** [comment]
@Cjballard awesome, thank you!  

@Guardiola31337 @devotaaabel will add tests to this PR tomorrow then will be good for 👀

**Comment by danesfeder** [comment]
Merge is causing issues with padding adjustments - this PR is no longer ready for review until these issues are fixed.

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/summary/SummaryBottomSheet.java [hunk 1]
```diff
@@ -177,6 +181,19 @@ private void bind() {
...
     arrivalTimeText = findViewById(R.id.arrivalTimeText);
     rerouteProgressBar = findViewById(R.id.rerouteProgressBar);
+    updateRouteOverviewImage();
+  }
+
+  private void updateRouteOverviewImage() {
+    ImageButton routeOverviewBtn = findViewById(R.id.routeOverviewBtn);
+    boolean isDarkThemeEnabled = ThemeSwitcher.isDarkThemeEnabled(getContext());
+    Drawable routeOverviewDrawable;
+    if (isDarkThemeEnabled) {
```
Comment (Guardiola31337) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/summary/SummaryBottomSheet.java]: Could you add a custom attribute to the nav theme so it's just added upon inflation and remove the `if else` check?

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/camera/SimpleCamera.java [hunk 2]
```diff
@@ -59,20 +62,36 @@ public double zoom(RouteInformation routeInformation) {
...
+      Point.fromLngLat(routeCoordinates.get(1).longitude(), routeCoordinates.get(1).latitude())
     );
   }
+
+  private List<Point> generateRouteCoordinates(DirectionsRoute route) {
+    if (route != null) {
+      LineString lineString = LineString.fromPolyline(route.geometry(), Constants.PRECISION_6);
+      return lineString.coordinates();
+    }
+    return null;
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/camera/SimpleCamera.java]: We should consider returning a [Null Object](https://en.wikipedia.org/wiki/Null_Object_pattern#Java) instead or even better try to getting rid of it at all (maybe returning an empty list?). If not, code becomes defensive and full of `null` checks.

👀 [Null References: The Billion Dollar Mistake](https://www.infoq.com/presentations/Null-References-The-Billion-Dollar-Mistake-Tony-Hoare)

Comment (Guardiola31337) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/camera/SimpleCamera.java]: Actually if the `route` in `setupLineStringAndBearing` is `null` L#85 will cause a NPE right?

Comment (Guardiola31337) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/camera/SimpleCamera.java]: In any case, guard clauses normally are at the beginning of the method representing the unexpected behavior.

Comment (Guardiola31337) [reply libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/camera/SimpleCamera.java]: Also saw the same exact code in `DynamicCameraTest.java` (`generateRouteCoordinates`). Should be abstracted and accessible to the implementation classes?

File: libandroid-navigation-ui/src/test/java/com/mapbox/services/android/navigation/ui/v5/camera/DynamicCameraTest.java [hunk 3]
```diff
@@ -145,6 +149,44 @@ public void onInformationFromLocationAndProgress_engineCreatesCorrectBearing() t
     assertEquals(100f, bearing, DELTA);
   }
 
+  @Test
+  public void onInformationFromRoute_engineCreatesOverviewPointList() throws Exception {
+    MapboxMap mapboxMap = mock(MapboxMap.class);
```
Comment (Guardiola31337) [inline libandroid-navigation-ui/src/test/java/com/mapbox/services/android/navigation/ui/v5/camera/DynamicCameraTest.java]: Minor thing - Could we extract some shared "arrangement" code into `private` methods and used it across the tests? This applies to `MapWaynameTest` as well.

**Comment by danesfeder** [comment]
@Guardiola31337 Updated - I skipped your comment on `MapWaynameTest` - the variables that are out of the builders are being used to provide context / used in the actual tests so they cannot be pushed into the builders.

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/camera/NavigationCamera.java [hunk 4]
```diff
@@ -246,6 +267,53 @@ public void onFinish() {
...
   }
 
+  private void animateCameraForRouteOverview(RouteInformation routeInformation, int[] padding) {
+    Camera cameraEngine = navigation.getCameraEngine();
+    List<Point> routePoints = cameraEngine.overview(routeInformation);
+    boolean invalidPoints = routePoints.isEmpty();
+    if (invalidPoints) {
+      return;
+    }
+    CameraUpdate resetUpdate = buildResetCameraUpdate();
```
Comment (Guardiola31337) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/camera/NavigationCamera.java]: https://github.com/mapbox/mapbox-navigation-android/pull/967#discussion_r195802121

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/camera/NavigationCamera.java [hunk 5]
```diff
@@ -246,6 +267,53 @@ public void onFinish() {
...
+    final CameraUpdate overviewUpdate = buildOverviewCameraUpdate(padding, routePoints);
+
+    mapboxMap.animateCamera(resetUpdate, 150, new MapboxMap.CancelableCallback() {
```
Comment (Guardiola31337) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/camera/NavigationCamera.java]: What about extracting the listeners into class properties?

File: libandroid-navigation-ui/src/test/java/com/mapbox/services/android/navigation/ui/v5/camera/DynamicCameraTest.java [hunk 6]
```diff
@@ -173,4 +214,12 @@ private DirectionsRoute buildDirectionsRoute() throws IOException {
     DirectionsResponse response = gson.fromJson(body, DirectionsResponse.class);
     return response.routes().get(0);
   }
+
+  private List<Point> generateRouteCoordinates(DirectionsRoute route) {
```
Comment (Guardiola31337) [inline libandroid-navigation-ui/src/test/java/com/mapbox/services/android/navigation/ui/v5/camera/DynamicCameraTest.java]: https://github.com/mapbox/mapbox-navigation-android/pull/967#discussion_r195812313

File: libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/utils/ViewUtils.java [hunk 7]
```diff
@@ -48,8 +49,12 @@ public static Bitmap loadBitmapFromView(View view) {
     return null;
   }
 
-  public static float dpToPx(Context context, int dp) {
+  public static int[] buildRouteOverviewPadding(Context context) {
```
Comment (Guardiola31337) [inline libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/utils/ViewUtils.java]: This doesn't sound like a `View` utility. It's only used in `NavigationView`. IMO, objects and methods should live/be where their context is. From my experience utility classes become a drawer easily and we should treat them carefully.

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/camera/SimpleCamera.java [hunk 8]
```diff
@@ -59,20 +63,36 @@ public double zoom(RouteInformation routeInformation) {
     return CAMERA_ZOOM;
   }
 
+  @Override
+  public List<Point> overview(RouteInformation routeInformation) {
```
Comment (Guardiola31337) [inline libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/camera/SimpleCamera.java]: IMO `overview` is getting too long.
What about extracting blocks of code into well-named private methods and simplifying conditional expressions somehow?

We should keep `public` methods as clean as possible so they read like pseudocode (i.e. no conditionals, `for`s, etc.).

**Pull request merged**